### ISSUE TITLE
[Feature Proposal] Multi-value Gauge

### ISSUE THREAD

**Issue by dflor003** [AppMetrics/AppMetrics#664]
# Proposal for a Multi-Value Gauge

I'm in need of a multi-value gauge for use with InfluxDB as described in #253 and looks like that issue was closed, so I'm going to start a new thread here to discuss what an implementation of a multi-value gauge could look like so that I can put together a PR to achieve this.

## Why is this useful?

Most time-series databases support some form of measurement with multiple fields reported at the same time. This is preferred over reporting each individual field as a separate metric because:

* Cross-series math can get quite complex in visualization tools like [Grafana](https://grafana.com/)
* If reporting each field as independent metrics, it becomes harder to correlate each individual data point to a single time frame
* When you have many fields, you end up shipping much more data to the time-series database and most of it is duplicated across the component metrics
* If your series tags have a high-cardinality, you end up increasing the number of series much more quickly if reporting as individual metrics

## Proposal

I propose adding a new type of gauge that supports multiple values that conform to a specific set of fields. To avoid breaking existing users of the Gauge metric, this would be a new metric called `MultiGauge` and it would include its own entry in the registry along with its own method of reporting values.

Let's spec this out by example from the perspective of the consumer of AppMetrics.

Assume we are trying to get a measurement like the following into the [InfluxDB line protocol](https://docs.influxdata.com/influxdb/cloud/reference/syntax/line-protocol/).

```
application.myapp__ram,tag1=value1 workingSet=209715200.0,private=134217728.0,virtual=52428800.0,maxWorkingSet=524288000.0 1556813561098000000
```

First, we need to declare a class that describes the set of fields we want to report:

```cs
public class RamUsage {
  public double WorkingSet { get; set; }
  public double Private { get; set; }
  public double Virtual { get; set; }
  public double MaxWorkingSet { get; set; }
}
```

Next, we declare the options using the `RamUsage` class as it's generic type:

```cs
public static readonly MultiGaugeOptions<RamUsage> RamMetric = 
  new MultiGaugeOptions<RamUsage> {
    Context = "Application.MyApp",
    Name = "Ram",
    MeasurementUnit = Unit.Bytes,
    ResetOnReporting = true,
  };
```

Behind the scenes, the metric would use reflection to extract the names of the fields from its generic type. It would need to statically perform the reflective calls and cache the results so that it does not impact performance.

Lastly, the consumer would report the metrics as follows:

```cs
public class SomeMetricReportingClass {
  private readonly IMetricsRoot _metrics;

  public SomeMetricReportingClass(IMetricsRoot metrics) {
    // For simplicity, lets assume we get IMetricsRoot from DI
    _metrics = metrics;
  }

  public void ReportMetrics() {
    // Gather the metrics
    var process = Process.GetCurrentProcess();
    var ramUsage = new RamUsage {
      MaxWorkingSet = process.MaxWorkingSet, 
      WorkingSet = process.WorkingSet64,
      Private = process.PrivateMemorySize64,
      Virtual = process. VirtualMemorySize64,
    };

    // Report the metrics
    _metrics.Measure.MultiGauge.SetValues(ramUsage);
  }
```

## Implementation

While I'm still familiarizing myself with the code, I think that this can be achieved by doing the following:

1. Create an `IMultiGauge<T>` interface where `T` is the type of the class from which the fields will be generated.
2. Similarly, create an `IMultiGaugeMetric<T>` interface that will extend `IMetricValueProvider<T>`.
3. Create a `MultiGaugeOptions<T>` class and build the logic to discover its properties and use them as fields. This should be done in the static constructor to ensure its only done once upon first use of the class and then cached from then on.
4. Add corresponding methods into `IMetricsRegistry` and `DefaultMetricsRegistry`.
5. Create the corresponding builders for `IMultiGauge`.
6. Create a `DefaultMultiGaugeManager` as well as an `IMeasureMultiGaugeMetrics` that will handle all multi-valued gauges.
7. Update `MetricSnapshotSerializer` to handle the multi-value gauge

**Note:** We'll probably need some kind of non-generic abstract base class for some of these to get the generics playing nicely in a non-generic context. Not sure about how that will look until I dig more into the code.

## Closing

I'm willing to put together a PR over the next few weeks to get this multi-value gauge implementation into AppMetrics. Tell me what you think of the plan and if there are any glaring issues with it. Let's continue discussions in the thread below.

**Comment by alhardy** [comment]
Thanks for the detail provided in the feature request.

The steps listed are pretty much what would be required to add the new metric type, it shouldn't be to complex following how others are implemented.

"use reflection to extract the names of the fields from its generic type": are there any concerns around performance here?

What would the non influxdb reporters would look like in terms of a multi-value gauge? Report as as single gauges where multi value is not supported?

**Comment by dflor003** [comment]
> "use reflection to extract the names of the fields from its generic type": are there any concerns around performance here?

Yeah, with reflection there definitely comes a performance cost. I'm thinking if we use a static constructor to reflectively load the public class properties and store them as static fields, it will be guaranteed to happen only once when the class is first initialized. This should mitigate the runtime performance penalty of reflection by moving it to a single time at startup.

> What would the non influxdb reporters would look like in terms of a multi-value gauge? Report as as single gauges where multi value is not supported?

Hmm... This warrants further research. As far as time-series DBs go, InfluxDB is the only one I have used. I see that, in addition to InfluxDB, the following are supported:

* DataDog
* Prometheus
* Graphite
* StatsD
* GrafanaCloudHostedMetrics

How are the metrics for multi-value metrics such as the histogram and timer formatted in these? For example, on the timer metric, I see it ships fields such as `last`, `min`, `max`, etc.

**Comment by alhardy** [comment]
"How are the metrics for multi-value metrics such as the histogram and timer formatted in these? For example, on the timer metric, I see it ships fields such as last, min, max, etc."

The get reported as individual metrics but have the same tags of course.

If a multi value gauge applied to at least a few TSDBs it be worthwhile adding, if its specific to influx it would be a little pointless and would suggest a influx specific implementation as opposed to AppMetrics

**Comment by alhardy** [comment]
Then again, I guess in some cases we could say a multi value gauge could be useful to group related metrics when recording without having to create a separate metric, even though if for example the specified reporter flushes them individually.

I think this sounds OK. I would like to see the support by other TSDBs though before just reporting them individually, and also test the performance in taking the reflection approach

**Comment by dflor003** [comment]
One idea for formatting a multi-value gauge in something that only supports single-values is to convert it automatically to one gauge per field. For example, if we had a gauge called `CpuUsage` with the fields `User`, `System`, and `ProcessorCount`, it could translate into 3 gauges: 'CpuUsage.User`, `CpuUsage.System`, `CpuUsage.ProcessorCount`.

Thoughts?

**Comment by alhardy** [comment]
Yep that was my thinking also

**Comment by dflor003** [comment]
Sounds good. Given that we have a plan for the TSDBs that don't support multi-value gauges. Should I go ahead and start on the changes? Any other concerns I should address before starting on that?

**Comment by alhardy** [comment]
Happy to go ahead. Could benchmarks be implemented for the implementation please
